File: src/services/inlayHints.ts
// Inlay hints: parameter names at call sites and inferred types for `var`.
// Offset-based so the unit tests stay position-free; the LSP server converts
// offsets to positions and applies the user's configuration.

import type { Checker } from "../compiler/checker.ts";
import { forEachChild } from "../compiler/parser.ts";
import {
  type CallExpression,
  type ForEachStatement,
  type LocalVariableDeclarationStatement,
  type Node,
  type Parameter,
  type SourceFile,
  SyntaxKind,
  type VariableDeclarator,
} from "../compiler/types.ts";

export interface InlayHintsSettings {
  /** Hints like `count:` before call arguments that are not plain variables. */
  parameterNames: boolean;
  /** Hints like `: String` after a `var` declaration's name. */
  varTypes: boolean;
}

export const DEFAULT_INLAY_HINTS: InlayHintsSettings = { parameterNames: true, varTypes: true };

export interface InlayHintEntry {
  readonly offset: number;
  readonly label: string;
  readonly kind: "parameter" | "type";
}

// A "self-explanatory" argument needs no name hint: a plain variable, `this`,
// or a field access whose final name already reads like the parameter.
function isSelfExplanatoryArgument(arg: Node): boolean {
  switch (arg.kind) {
    case SyntaxKind.Identifier:
    case SyntaxKind.ThisExpression:
    case SyntaxKind.PropertyAccessExpression:
      return true;
    default:
      return false;
  }
}

export function getInlayHints(
  checker: Checker,
  sourceFile: SourceFile,
  startOffset: number,
  endOffset: number,
  settings: InlayHintsSettings = DEFAULT_INLAY_HINTS,
): InlayHintEntry[] {
  const hints: InlayHintEntry[] = [];

  const collectCallHints = (call: CallExpression): void => {
    if (call.arguments.length === 0) return;
    const decl = checker.resolveCall(call);
    if (!decl) return;
    call.arguments.forEach((arg, i) => {
      if (arg.end < startOffset || arg.pos > endOffset) return;
      if (isSelfExplanatoryArgument(arg)) return;
      const param = decl.parameters[i] as Parameter | undefined;
      const name = param?.name?.text;
      if (!name) return;
      if (param.isVarArgs) {
        // Only the first argument of the variable-arity tail gets a `...name:`.
        if (i !== decl.parameters.length - 1) return;
        hints.push({ offset: skipToStart(arg), label: `...${name}:`, kind: "parameter" });
        return;
      }
      hints.push({ offset: skipToStart(arg), label: `${name}:`, kind: "parameter" });
    });
  };

  // `pos` includes leading trivia; the hint goes right before the first token.
  const skipToStart = (node: Node): number => {
    let pos = node.pos;
    const text = sourceFile.text;
    while (pos < node.end && /\s/.test(text[pos]!)) pos++;
    return pos;
  };

  const varTypeHint = (declarator: VariableDeclarator | Parameter): void => {
    const name = (declarator as { name?: { text: string; end: number } }).name;
    const symbol = (declarator as Node).symbol;
    if (!name || !symbol) return;
    if (name.end < startOffset || name.end > endOffset) return;
    const type = checker.typeStringOfSymbol(symbol);
    if (type === "<error>" || type === "var" || type === "") return;
    hints.push({ offset: name.end, label: `: ${type}`, kind: "type" });
  };

  const visit = (node: Node): void => {
    if (node.end < startOffset || node.pos > endOffset) return; // outside the range
    if (settings.parameterNames && node.kind === SyntaxKind.CallExpression) {
      collectCallHints(node as CallExpression);
    }
    if (settings.varTypes && node.kind === SyntaxKind.LocalVariableDeclarationStatement) {
      const s = node as LocalVariableDeclarationStatement;
      if (s.type.kind === SyntaxKind.VarType) {
        for (const d of s.declarators) varTypeHint(d as VariableDeclarator);
      }
    }
    if (settings.varTypes && node.kind === SyntaxKind.ForEachStatement) {
      const parameter = (node as ForEachStatement).parameter;
      if (parameter.type.kind === SyntaxKind.VarType) varTypeHint(parameter);
    }
    forEachChild(node, child => {
      visit(child);
      return undefined;
    });
  };
  visit(sourceFile);
  hints.sort((a, b) => a.offset - b.offset);
  return hints;
}

File: togo/internal/services/inlay_hints.go
package services

// Inlay hints: parameter names at call sites and inferred types for `var`.
// Offset-based so the unit tests stay position-free; the LSP server converts
// offsets to positions and applies the user's configuration.
// Port of src/services/inlayHints.ts.

import (
	"sort"

	"github.com/nikeee/cappu/internal/compiler"
)

// InlayHintsSettings toggles each hint family.
type InlayHintsSettings struct {
	ParameterNames bool // hints like `count:` before call arguments
	VarTypes       bool // hints like `: String` after a `var` declaration's name
}

// DefaultInlayHints enables both families.
var DefaultInlayHints = InlayHintsSettings{ParameterNames: true, VarTypes: true}

// InlayHintEntry is one hint at an offset.
type InlayHintEntry struct {
	Offset int
	Label  string
	Kind   string // "parameter" | "type"
}

func isSelfExplanatoryArgument(arg *compiler.Node) bool {
	switch arg.Kind {
	case compiler.Identifier, compiler.ThisExpression, compiler.PropertyAccessExpression:
		return true
	default:
		return false
	}
}

func isSpace(b byte) bool {
	return b == ' ' || b == '\t' || b == '\n' || b == '\r' || b == '\v' || b == '\f'
}

// GetInlayHints collects parameter-name and var-type hints in [startOffset, endOffset].
func GetInlayHints(checker *compiler.Checker, sourceFile *compiler.Node, startOffset, endOffset int, settings InlayHintsSettings) []InlayHintEntry {
	text := sourceFile.AsSourceFile().Text
	var hints []InlayHintEntry

	skipToStart := func(node *compiler.Node) int {
		pos := node.Pos
		for pos < node.End && pos < len(text) && isSpace(text[pos]) {
			pos++
		}
		return pos
	}

	varTypeHint := func(declarator *compiler.Node) {
		name := declName(declarator)
		symbol := declarator.Symbol
		if name == nil || symbol == nil {
			return
		}
		if name.End < startOffset || name.End > endOffset {
			return
		}
		typ := checker.TypeStringOfSymbol(symbol)
		if typ == "<error>" || typ == "var" || typ == "" {
			return
		}
		hints = append(hints, InlayHintEntry{Offset: name.End, Label: ": " + typ, Kind: "type"})
	}

	collectCallHints := func(call *compiler.Node) {
		args := call.AsCallExpression().Arguments
		if args == nil || args.Len() == 0 {
			return
		}
		decl := checker.ResolveCall(call)
		if decl == nil {
			return
		}
		params := decl.AsMethodDeclaration().Parameters
		for i, arg := range args.Nodes {
			if arg.End < startOffset || arg.Pos > endOffset {
				continue
			}
			if isSelfExplanatoryArgument(arg) {
				continue
			}
			if i >= params.Len() {
				continue
			}
			param := params.Nodes[i].AsParameter()
			if param.Name == nil {
				continue
			}
			name := param.Name.AsIdentifier().Text
			if param.IsVarArgs {
				if i != params.Len()-1 {
					continue
				}
				hints = append(hints, InlayHintEntry{Offset: skipToStart(arg), Label: "..." + name + ":", Kind: "parameter"})
				continue
			}
			hints = append(hints, InlayHintEntry{Offset: skipToStart(arg), Label: name + ":", Kind: "parameter"})
		}
	}

	var visit compiler.Visitor
	visit = func(node *compiler.Node) bool {
		if node.End < startOffset || node.Pos > endOffset {
			return false
		}
		if settings.ParameterNames && node.Kind == compiler.CallExpression {
			collectCallHints(node)
		}
		if settings.VarTypes && node.Kind == compiler.LocalVariableDeclarationStatement {
			s := node.AsLocalVariableDeclarationStatement()
			if s.Type.Kind == compiler.VarType {
				for _, d := range s.Declarators.Nodes {
					varTypeHint(d)
				}
			}
		}
		if settings.VarTypes && node.Kind == compiler.ForEachStatement {
			parameter := node.AsForEachStatement().Parameter
			if parameter.AsParameter().Type.Kind == compiler.VarType {
				varTypeHint(parameter)
			}
		}
		node.ForEachChild(visit)
		return false
	}
	visit(sourceFile)
	sort.SliceStable(hints, func(i, j int) bool { return hints[i].Offset < hints[j].Offset })
	return hints
}
